src/mlpack/methods/ann/layer/gru_impl.hpp

```diff
@@ -0,0 +1,307 @@
+
+  // Delta rt.
+  arma::mat d_rt = boost::apply_visitor(deltaVisitor, outputHidden2GateModule) %
+      *backIterator;
+
+  // Delta of forget gate.
+  boost::apply_visitor(BackwardVisitor(std::move(boost::apply_visitor(
+      outputParameterVisitor, forgetGateModule)), std::move(d_rt),
+      std::move(boost::apply_visitor(deltaVisitor, forgetGateModule))),
+      forgetGateModule);
+
+  // Put delta zt.
+  prevError.submat(0, 0, 1 * outSize - 1, 0) = boost::apply_visitor(
+      deltaVisitor, inputGateModule);
+
+  // Put delta rt.
+  prevError.submat(1 * outSize, 0, 2 * outSize - 1, 0) = boost::apply_visitor(
+      deltaVisitor, forgetGateModule);
+
+  // Put delta ot.
+  prevError.submat(2 * outSize, 0, 3 * outSize - 1, 0) = boost::apply_visitor(
+      deltaVisitor, hiddenStateModule);
+
+  // Get delta ht - 1 for input gate and forget gate.
+  boost::apply_visitor(BackwardVisitor(std::move(boost::apply_visitor(
+      outputParameterVisitor, input2GateModule)), 
+      std::move(prevError.submat(0, 0, 2 * outSize - 1, 0)),
+      std::move(boost::apply_visitor(deltaVisitor, output2GateModule))),
+      output2GateModule);   
+
+  // Add delta ht - 1 from hidden state.
+  boost::apply_visitor(deltaVisitor, output2GateModule) += 
+      boost::apply_visitor(deltaVisitor, outputHidden2GateModule) %
+      boost::apply_visitor(outputParameterVisitor, forgetGateModule);
+
+  // Add delta ht - 1 from ht.
+  boost::apply_visitor(deltaVisitor, output2GateModule) += gy %
+      boost::apply_visitor(outputParameterVisitor, inputGateModule);
+
+  // Get delta input.
+  boost::apply_visitor(BackwardVisitor(std::move(boost::apply_visitor(
+      outputParameterVisitor, input2GateModule)), std::move(prevError),
+      std::move(boost::apply_visitor(deltaVisitor, input2GateModule))),
+      input2GateModule);
+
+  backwardStep++;
+  backIterator--;
+
+  g = boost::apply_visitor(deltaVisitor, input2GateModule);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void GRU<InputDataType, OutputDataType>::Gradient(
+    arma::Mat<eT>&& input,
+    arma::Mat<eT>&& /* error */,
+    arma::Mat<eT>&& /* gradient */)
+{
+  if (gradIterator == outParameter.end())
+  {
+    gradIterator = --(--outParameter.end());
+  }
+
+  boost::apply_visitor(GradientVisitor(std::move(input), std::move(prevError)),
+      input2GateModule);
+
+  boost::apply_visitor(GradientVisitor(
+      std::move(*gradIterator),
+      std::move(prevError.submat(0, 0, 2 * outSize - 1, 0))), 
+      output2GateModule);
+
+  boost::apply_visitor(GradientVisitor(
+      std::move(*gradIterator),
+      std::move(prevError.submat(2 * outSize, 0, 3 * outSize - 1, 0))), 
+      outputHidden2GateModule);
+
+  gradIterator--;
+}
+
+template<typename InputDataType, typename OutputDataType>
+void GRU<InputDataType, OutputDataType>::ResetCell()
+{
+  outParameter.clear();
+  outParameter.push_back(arma::zeros<arma::mat>(outSize, 1));
```

Comment:
I am sorry I did not get that. Yo mean we clear the list till the first element?